src/api.py

```diff
@@ -0,0 +1,239 @@
+    return listfiles(upload_folder(), ext_list=['.json'])
+
+def xml_files():
+    return listfiles(upload_folder(), ext_list=['.xml'])
+
+#
+# api
+#
+
+def search_xml():
+    "GET /xml"
+    return xml_files()[0] # just the basenames
+
+def search_article_json():
+    return article_json_files()[0] # just the basenames
+
+def get_article_json(filename):
+    basenames, full_paths = article_json_files()
+    if not filename:
+        return basenames
+    http_ensure(filename in basenames, "requested file not found", 404)
+    idx = dict(zip(basenames, full_paths))
+    return json.load(open(idx[filename], 'r'))
+
+def post_xml():
+    "upload jats xml, generate xml, validate, send to lax as a dry run"
+    http_ensure('xml' in request.files, "xml file required", 404)
+
+    try:
+        override = scraper.deserialize_overrides(request.form.getlist('override'))
+    except ValueError as err:
+        sio = StringIO()
+        traceback.print_exc(file=sio)
+        return {
+            'status': conf.ERROR,
+            'code': 'bad-overrides',
+            'message': 'an error occurred attempting to parse your given overrides.',
+            'trace': sio.getvalue()
+        }
+
+    # upload
+    try:
+        xml = request.files['xml']
+        filename = xml.filename # todo: sanitize this. assumes a name like 'elife-00000-v1.xml'
+        http_ensure(os.path.splitext(filename)[1] == '.xml', "file doesn't look like xml")
+        path = join(upload_folder(), filename)
+        xml.save(path)
+    except Exception as err:
+        sio = StringIO()
+        traceback.print_exc(file=sio)
+        return {
+            'status': conf.ERROR,
+            'code': 'error-uploading-xml',
+            'message': 'an error occured uploading the article xml to be processed',
+            'trace': sio.getvalue(),
+        }, 400 # everything is always the client's fault.
+
+    # generate
+    try:
+        article_json = scraper.main(path, {'override': override})
+        json_filename = filename + '.json'
+        json_path = join(upload_folder(), json_filename)
+        open(json_path, 'w').write(article_json)
+    except Exception as err:
+        sio = StringIO()
+        traceback.print_exc(file=sio)
+        return {
+            'status': conf.ERROR,
+            'code': 'error-scraping-xml',
+            'message': str(err),
+            'trace': sio.getvalue()
+        }, 400
+
+    # validate
+    try:
+        ajson_validate.main(open(json_path, 'r'))
+
+    except jsonschema.ValidationError as err:
+        return {
+            'status': 'invalid',
+            'code': 'invalid-article-json',
+            'message': 'the generated article-json failed validation',
+            'trace': str(err), # todo: any good?
+        }, 400
+    except Exception as err:
+        sio = StringIO()
+        traceback.print_exc(file=sio)
+        return {
+            'status': conf.ERROR,
+            'code': 'error-validating-article-json',
```

Comment:
codes are not documented anywhere like that are in lax. they probably should be and an enum in the schema would be helpful.